duckduckgo/iOS
# Autofill password generation support for iOS

Task/Issue URL: https://app.asana.com/0/72649045549333/1204452429715454/f
Tech Design URL:
CC:

**Description**:
Adds support for autofill password generation on iOS



**Steps to test this PR**:

Notes:
1. Modify `FeatureFlag.swift` and set `.autofillPasswordGeneration` to `.internalOnly`
2. Visit Browsing Menu -> Logins and delete logins between tests

_Use generated password and complete form_
1. Visit https://fill.dev/form/registration-username
2. Enter username
3. Tap on password field
4. Verify prompt appears offering a generated password
5. Accept to use it
6. Verify it populates the password and confirm password fields
7. Submit form
8. Verify toast shown confirming login saved
9. Tap on toast's `View` button; confirm you see the new login

_Use generated password but abandon form_
1. Visit https://fill.dev/form/registration-username
2. Enter username
3. Tap on password field
4. Verify prompt appears offering a generated password
5. Accept to use it
6. Verify it populates the password and confirm password fields
7. Don't submit form; instead, reload the page or navigate away
8. Visit  the Logins screen and verify the login with the username and generated password has been saved

_Use generated password then edit the generated password_
1. Visit https://fill.dev/form/registration-username
2. Enter username
3. Tap on password field
4. Verify prompt appears offering a generated password
5. Accept to use it
6. Manually edit the password and confirm password fields
7. Submit the form
8. Verify you are prompted if you want to save the login
9. Tap "not now" and verify no login was saved

_Decline to use generated password_
1. Visit https://fill.dev/form/registration-username
2. Enter username
3. Tap on password field
4. Verify prompt appears offering a generated password
5. Tap on "Not Now"; enter your own password in both fields
6. Submit form
7. Verify you are prompted if you want to save the login

_Copy generated password_
1. Visit https://fill.dev/form/registration-username
2. Tap on password field
3. Tap on the copy icon to the right of the password
4. Verify it copied successfully to the clipboard by pasting it somewhere

_Username updated later_
1. Visit https://fill.dev/form/registration-username
2. Don't enter username
3. Tap password and accept to use generated password
4. Then go enter a username and submit the form
5. Verify in Logins screen that the saved credential has both the username and password

_Confirm fireproof dialog does not overlay login saved toast_
1. In Settings > Fireproof > Fireproof Site toggle the `Ask to Fireproof sites on sign-in` to ON
2. Visit https://fill.dev/form/registration-username
3. Enter username
4. Tap on password field
5. Verify prompt appears offering a generated password
6. Accept to use it
7. Verify it populates the password and confirm password fields
8. Submit form
9. Verify toast shown confirming login saved 
10. Verify that prompt to fireproof displays after the toast has disappeared

_Disable toggle in autofill settings_
1. Go to the Logins screen and disable toggle for `Save and Autofill Logins`
2. Visit https://fill.dev/form/registration-username
3. Tap on password field; verify you are NOT prompted to use a generated password

_Smoke test existing save login flows_
1. Visit Logins screen and re-enable toggle for `Save and Autofill Logins`
2. Visit https://fill.dev/form/login-simple
3. Enter a username and password
4. Verify the prompt to save the login is presented
5. Save the login
6. Confirm the login has saved in the Logins screen

_Test remote config with config disabled_
1. Modify `FeatureFlag.swift` `.autofillPasswordGeneration` back to `.remoteReleasable(.subfeature(AutofillSubfeature.autofillPasswordGeneration))`
2. Visit https://fill.dev/form/registration-username
3. Tap on password field; verify you are NOT prompted to use a generated password



**Copy Testing**:


**Orientation Testing**:


**Device Testing**:


**OS Testing**:


**Theme Testing**:


---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
